### PR TITLE
cbify: support adf reduction

### PR DESCRIPTION
**Pull request by albietz**
Add support for the adf contextual bandit reduction from cbify: for each (single line) multiclass example, this generates multiline examples (one per action, then one empty example) by shifting feature indicies for each action, and reduces to cb_explore_adf instead of cb_explore.

### CODE REVIEW

**Comment by arielf** [comment]
I think this PR is good.  The AppVeyor failure seems unrelated.

File: vowpalwabbit/cbify.cc [hunk 1]
```diff
@@ -64,6 +73,42 @@ void finish(cbify& data)
   delete_it(data.mwt_explorer);
   delete_it(data.recorder);
   data.a_s.delete_v();
+  if (data.use_adf)
+  { for (size_t a = 0; a < data.adf_data.num_actions; ++a)
+    { VW::dealloc_example(CB::cb_label.delete_label, data.adf_data.ecs[a]);
```
Comment (JohnLangford) [inline vowpalwabbit/cbify.cc]: It's computationally inelegant to dealloc a large number of examples often.   Can you tweak the code to just reuse the examples instead?  That's what the rest of VW does.

File: vowpalwabbit/cbify.cc [hunk 2]
```diff
@@ -107,14 +201,20 @@ base_learner* cbify_setup(vw& all)
...
+  data.all = &all;
 
-  if (count(all.args.begin(), all.args.end(),"--cb_explore") == 0)
+  if (data.use_adf)
+  { init_adf_data(data, num_actions);
+  }
+
+  if (count(all.args.begin(), all.args.end(),"--cb_explore") == 0 && !data.use_adf)
   { all.args.push_back("--cb_explore");
     stringstream ss;
```
Comment (JohnLangford) [inline vowpalwabbit/cbify.cc]: I expected a pushback of --cb_explore_adf here?

File: vowpalwabbit/cbify.cc (hunk 1 above)
Comment (albietz) [reply vowpalwabbit/cbify.cc]: Doesn't the `finish` function here just get called on termination (through `set_finish`)? I am allocating the examples once on setup (see `init_adf_data`), reusing them on every example, and this call deallocs them.

**Pull request merged**